blueprints/cloud-operations/network-dashboard/cloud-function/metrics/subnets.py

```diff
@@ -214,9 +238,90 @@ def compute_subnet_utilization(config, all_subnets_dict):
     # PSA Range for Cloud SQL, MemoryStore, etc.
     elif purpose == "VPC_PEERING":
-      # TODO: PSA range to be handled later
-      # print("PSA range to be handled later:", address, prefixLength, network_name)
-      continue
+      ip_range = f"{address}/{int(prefixLength)}"
+      net = ipaddress.ip_network(ip_range)
+      # Note that 4 IP addresses are reserved by GCP in all subnets
+      # Source: https://cloud.google.com/vpc/docs/subnets#reserved_ip_addresses_in_every_subnet
+      total_ip_addresses = int(net.num_addresses) - 4
+      all_subnets_dict[project_id][f"psa/{address_name}"] = {
+          'name': f"psa/{address_name}",
+          'region': subnet_region,
+          'ip_cidr_range': ip_range,
+          'total_ip_addresses': total_ip_addresses,
+          'used_ip_addresses': 0,
+          'network_name': network_name
+      }
+
+
+def compute_subnet_utilization_redis(config, read_mask, all_subnets_dict):
+  '''
+    Counts Redis (Memorystore) instances using private IPs in the different subnets.
+      Parameters:
+        config (dict): Dict containing config like clients and limits
+        read_mask (FieldMask): read_mask to get additional metadata from Cloud Asset Inventory
+        all_subnets_dict (dict): Dict containing the information for each subnets in the GCP organization
+      Returns:
+        None
+  '''
+  response_redis = config["clients"]["asset_client"].search_all_resources(
+      request={
+          "scope": f"organizations/{config['organization']}",
+          "asset_types": ["redis.googleapis.com/Instance"],
+          "read_mask": read_mask,
+          "page_size": config["page_size"],
+      })
+
+  for asset in response_redis:
+    ip_range = ""
+    connect_mode = ""
+    network_name = ""
+    project_id = ""
+    region = ""
+    for versioned in asset.versioned_resources:
+      for field_name, field_value in versioned.resource.items():
+        if field_name == 'locationId':
+          region = field_value[0:-2]
+        if field_name == 'authorizedNetwork':
+          network_name = field_value.split('/')[-1]
+          project_id = field_value.split('/')[1]
+        if field_name == 'reservedIpRange':
+          ip_range = field_value
+        if field_name == 'connectMode':
+          connect_mode = field_value
+
+    # Only handling PSA for Redis for now
+    if connect_mode == "PRIVATE_SERVICE_ACCESS":
+      # Reddis instance asset doesn't contain the subnet information in Asset Inventory
+      # We need to find the correct subnet with IP address matching
+      redis_ip_range = ipaddress.ip_network(ip_range)
+      for subnet_key, subnet_dict in all_subnets_dict[project_id].items():
+        if subnet_dict["network_name"] == network_name:
+          if redis_ip_range.overlaps(
+              ipaddress.ip_network(subnet_dict['ip_cidr_range'])):
+            all_subnets_dict[project_id][subnet_key][
+                'used_ip_addresses'] += redis_ip_range.num_addresses
```

Comment:
Do we really know if Redis will consume all of these IP Addresses or just needs them in case of?

Reply:
I don't think Redis will consume all these IPs but they are reserved and cannot be used for other services. So for now, I will assume they are not available (even if that is a bit conservative).